Errors are not something scary.
Every program that runs in production is expected to manage errors on a constant basis. 
Our job as devs is to handle the errors gracefully.
In a way that aligns with our user's expectations. 

### Error are NOT bugs.
Somthin u don't expect happens in ur code -> raise own exceptions.

For example, 
If someone passes some bad inputs to a function we write, 
we should not be afraid to raise an exception to let them know they what's wrong.

An error or exception is raised when something bad happens, 
but as long as our code handlees it as users expect it to... it's not a bug. 

---
A bug is when code behaves in ways our users don't expect to.

For example,
If player tries forge an iron sword out bronze metal, raise exception and display error msg.
However, If a player can forge the iron sword out of bronze, 
that is a bug because that's against the rules of the game.

As a rule of thumb. 
You don't wanna handle your own exceptions within the same function block.
Instead, the caller should handle any potential error 
by wrapping the function call within a try/except block.

#### Syntax for raising exceptions
``` python
raise Exception('somethin bad happen')
```

## Review
There are many types of exceptions.
Specific exceptions are built into language like `IndexError` or `ZeroDivisionError`,
and almost all Exceptions count as the parent `Exception` type.
What differentiates exceptions are their types, not their string descriptions

This is a list of built-in exceptions -> [Exceptions List](https://docs.python.org/3/library/exceptions.html) 

Software apps ain't perfect.
User input and network connectivity are far from predictable.
Despite intensive debugging and unit testing, apps will still have failure cases.
- loss of network
- missing database rows, 
- out of memory issues, 
- unexpected user inputs 
can all prevent an app from performing correctly.
It is our job to catch and handle any and all exceptions gracefully.
When u are able to detect that something is amiss, 
U should raise the errors urself,  in addition to what python might raise

`raise` keyword triggers an exception in Python
Good code handles errors.



---
[[12.2_Except_Review]]
[[12.4_Raise_vs_exception]]